pkg/clustermesh/mcsapi/serviceimport_controller.go

```
@@ -0,0 +1,542 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	conditionTypeReady = "Ready"
+)
+
+// mcsAPIServiceImportReconciler is a controller that automatically creates
+// ServiceImport from ServiceExport (and their corresponding Services) from
+// remote clusters and the local cluster. It also handles ServiceExport status
+// since we have all the info here to do so.
+type mcsAPIServiceImportReconciler struct {
+	client.Client
+	Logger logrus.FieldLogger
+
+	clusterName                string
+	globalServiceExports       *operator.GlobalServiceExportCache
+	remoteClusterServiceSource *remoteClusterServiceExportSource
+}
+
+func newMCSAPIServiceImportReconciler(mgr ctrl.Manager, logger logrus.FieldLogger, clusterName string, globalServiceExports *operator.GlobalServiceExportCache, remoteClusterServiceSource *remoteClusterServiceExportSource) *mcsAPIServiceImportReconciler {
+	return &mcsAPIServiceImportReconciler{
+		Client:                     mgr.GetClient(),
+		Logger:                     logger,
+		clusterName:                clusterName,
+		globalServiceExports:       globalServiceExports,
+		remoteClusterServiceSource: remoteClusterServiceSource,
+	}
+}
+
+func (r *mcsAPIServiceImportReconciler) getSvcExport(ctx context.Context, req ctrl.Request) (*mcsapiv1alpha1.ServiceExport, error) {
+	var svcExport mcsapiv1alpha1.ServiceExport
+	if err := r.Client.Get(ctx, req.NamespacedName, &svcExport); err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+	return &svcExport, nil
+}
+
+func (r *mcsAPIServiceImportReconciler) getSvcImportBase(ctx context.Context, req ctrl.Request) (*mcsapiv1alpha1.ServiceImport, bool, error) {
+	var svcImport mcsapiv1alpha1.ServiceImport
+	if err := r.Client.Get(ctx, req.NamespacedName, &svcImport); err != nil {
+		if k8sApiErrors.IsNotFound(err) {
+			return &mcsapiv1alpha1.ServiceImport{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      req.Name,
+					Namespace: req.Namespace,
+				},
+			}, false, nil
+		}
```

Should return the error if different from NotFound.